File: agent_service_agreements/agreement.py
"""Agreement class for Agent Service Agreements.

Machine-readable service contracts with quality criteria, SLO definitions,
payment terms, JSON serialization, and SHA-256 agreement hashing.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from .schema import (
    ASA_VERSION,
    AGREEMENT_STATUSES,
    EscrowConfig,
    Identity,
    QualityCriteria,
    ServiceSpec,
    VerificationConfig,
    _hash_dict,
    _now_iso,
    _uuid,
)


@dataclass
class Agreement:
    """A machine-readable service agreement between a client and provider.

    Lifecycle: PROPOSED -> NEGOTIATING -> ACTIVE -> DELIVERED -> VERIFIED -> CLOSED
    """

    # Core identifiers
    agreement_id: str = ""
    asa_version: str = ASA_VERSION
    created_at: str = ""
    expires_at: str = ""
    status: str = "proposed"

    # Parties
    client: Optional[Identity] = None
    provider: Optional[Identity] = None
    evaluator: Optional[Identity] = None
    evaluator_type: str = "agent_as_judge"

    # Service definition
    service: Optional[ServiceSpec] = None

    # Quality criteria
    quality_criteria: Optional[QualityCriteria] = None

    # Verification configuration
    verification: Optional[VerificationConfig] = None

    # Escrow configuration
    escrow: Optional[EscrowConfig] = None

    # Dispute configuration
    dispute_protocol: str = "ajp"
    dispute_auto_file_threshold: float = 60.0

    # Signatures
    client_signature: str = ""
    provider_signature: str = ""

    # Deliverable (set when delivered)
    deliverable_hash: str = ""
    delivered_at: str = ""

    # Hash
    agreement_hash: str = ""

    def __post_init__(self) -> None:
        if not self.agreement_id:
            self.agreement_id = f"asa-{_uuid()[:12]}"
        if not self.created_at:
            self.created_at = _now_iso()

    def compute_hash(self) -> str:
        """Compute SHA-256 hash of the agreement's canonical form."""
        d = self._canonical_dict()
        self.agreement_hash = _hash_dict(d)
        return self.agreement_hash

    def _canonical_dict(self) -> Dict[str, Any]:
        """Fields included in the agreement hash (excludes mutable state)."""
        d: Dict[str, Any] = {
            "agreement_id": self.agreement_id,
            "asa_version": self.asa_version,
            "created_at": self.created_at,
        }
        if self.client:
            d["client"] = self.client.to_dict()
        if self.provider:
            d["provider"] = self.provider.to_dict()
        if self.service:
            d["service"] = self.service.to_dict()
        if self.quality_criteria:
            d["quality_criteria"] = self.quality_criteria.to_dict()
        if self.escrow:
            d["escrow"] = self.escrow.to_dict()
        return d

    def validate(self) -> List[str]:
        """Validate the agreement. Returns list of error messages (empty = valid)."""
        errors: List[str] = []
        if not self.client:
            errors.append("Agreement must have a client identity")
        if not self.provider:
            errors.append("Agreement must have a provider identity")
        if not self.service:
            errors.append("Agreement must specify a service")
        if not self.quality_criteria:
            errors.append("Agreement must have quality criteria")
        elif self.quality_criteria:
            if not self.quality_criteria.dimensions:
                errors.append("Quality criteria must include at least one dimension")
            else:
                total_weight = sum(d.weight for d in self.quality_criteria.dimensions)
                if abs(total_weight - 1.0) > 0.05:
                    errors.append(
                        f"Dimension weights should sum to ~1.0, got {total_weight:.2f}"
                    )
        if self.status not in AGREEMENT_STATUSES:
            errors.append(f"Invalid status: {self.status}")
        return errors

    def is_valid(self) -> bool:
        return len(self.validate()) == 0

    # -- Lifecycle transitions --

    def sign(self, party: str, signature: str) -> None:
        """Sign the agreement. Both parties must sign to activate."""
        if party == "client":
            self.client_signature = signature
        elif party == "provider":
            self.provider_signature = signature
        if self.client_signature and self.provider_signature:
            self.status = "active"
            self.compute_hash()

    def deliver(self, deliverable_hash: str) -> None:
        """Mark agreement as delivered with content hash."""
        if self.status != "active":
            raise ValueError(f"Cannot deliver in status '{self.status}', must be 'active'")
        self.deliverable_hash = deliverable_hash
        self.delivered_at = _now_iso()
        self.status = "delivered"

    def mark_verified(self, passed: bool) -> None:
        """Transition to verified state."""
        if self.status != "delivered":
            raise ValueError(f"Cannot verify in status '{self.status}', must be 'delivered'")
        self.status = "verified"

    def close(self) -> None:
        """Close the agreement."""
        if self.status not in ("verified", "disputed"):
            raise ValueError(f"Cannot close in status '{self.status}'")
        self.status = "closed"

    def dispute(self) -> None:
        """Transition to disputed state."""
        if self.status not in ("verified", "delivered"):
            raise ValueError(f"Cannot dispute in status '{self.status}'")
        self.status = "disputed"

    def expire(self) -> None:
        """Expire the agreement (timeout)."""
        self.status = "expired"

    def reject(self) -> None:
        """Reject during negotiation."""
        if self.status not in ("proposed", "negotiating"):
            raise ValueError(f"Cannot reject in status '{self.status}'")
        self.status = "rejected"

    # -- Serialization --

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "asa_version": self.asa_version,
            "agreement_id": self.agreement_id,
            "created_at": self.created_at,
            "status": self.status,
        }
        if self.expires_at:
            d["expires_at"] = self.expires_at

        parties: Dict[str, Any] = {}
        if self.client:
            parties["client"] = self.client.to_dict()
        if self.provider:
            parties["provider"] = self.provider.to_dict()
        if self.evaluator:
            parties["evaluator"] = self.evaluator.to_dict()
            parties["evaluator"]["type"] = self.evaluator_type
        if parties:
            d["parties"] = parties

        if self.service:
            d["service"] = self.service.to_dict()
        if self.quality_criteria:
            d["quality_criteria"] = self.quality_criteria.to_dict()
        if self.verification:
            d["verification"] = self.verification.to_dict()
        if self.escrow:
            d["escrow"] = self.escrow.to_dict()

        d["dispute"] = {
            "protocol": self.dispute_protocol,
            "auto_file_threshold": self.dispute_auto_file_threshold,
        }

        sigs: Dict[str, str] = {}
        if self.client_signature:
            sigs["client"] = self.client_signature
        if self.provider_signature:
            sigs["provider"] = self.provider_signature
        if sigs:
            d["signatures"] = sigs

        if self.deliverable_hash:
            d["deliverable_hash"] = self.deliverable_hash
        if self.delivered_at:
            d["delivered_at"] = self.delivered_at
        if self.agreement_hash:
            d["agreement_hash"] = self.agreement_hash

        return d

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "Agreement":
        parties = d.get("parties", {})
        client_d = parties.get("client")
        provider_d = parties.get("provider")
        evaluator_d = parties.get("evaluator")
        sigs = d.get("signatures", {})
        dispute = d.get("dispute", {})

        evaluator_type = "agent_as_judge"
        if evaluator_d and "type" in evaluator_d:
            evaluator_type = evaluator_d.pop("type", "agent_as_judge")

        return cls(
            agreement_id=d.get("agreement_id", ""),
            asa_version=d.get("asa_version", ASA_VERSION),
            created_at=d.get("created_at", ""),
            expires_at=d.get("expires_at", ""),
            status=d.get("status", "proposed"),
            client=Identity.from_dict(client_d) if client_d else None,
            provider=Identity.from_dict(provider_d) if provider_d else None,
            evaluator=Identity.from_dict(evaluator_d) if evaluator_d else None,
            evaluator_type=evaluator_type,
            service=ServiceSpec.from_dict(d["service"]) if d.get("service") else None,
            quality_criteria=QualityCriteria.from_dict(d["quality_criteria"]) if d.get("quality_criteria") else None,
            verification=VerificationConfig.from_dict(d["verification"]) if d.get("verification") else None,
            escrow=EscrowConfig.from_dict(d["escrow"]) if d.get("escrow") else None,
            dispute_protocol=dispute.get("protocol", "ajp"),
            dispute_auto_file_threshold=dispute.get("auto_file_threshold", 60.0),
            client_signature=sigs.get("client", ""),
            provider_signature=sigs.get("provider", ""),
            deliverable_hash=d.get("deliverable_hash", ""),
            delivered_at=d.get("delivered_at", ""),
            agreement_hash=d.get("agreement_hash", ""),
        )

    def to_json(self, indent: int = 2) -> str:
        """Serialize to JSON string."""
        import json
        return json.dumps(self.to_dict(), indent=indent, ensure_ascii=False)

    @classmethod
    def from_json(cls, s: str) -> "Agreement":
        """Deserialize from JSON string."""
        import json
        return cls.from_dict(json.loads(s))

File: typescript/src/agreement.ts
import {
  ASA_VERSION,
  AGREEMENT_STATUSES,
  AgreementStatus,
  EscrowConfig,
  Identity,
  QualityCriteria,
  ServiceSpec,
  VerificationConfig,
  hashDict,
  nowIso,
  uuid,
} from "./schema";

export interface AgreementDict {
  asa_version: string;
  agreement_id: string;
  created_at: string;
  status: string;
  expires_at?: string;
  parties?: Record<string, unknown>;
  service?: Record<string, unknown>;
  quality_criteria?: Record<string, unknown>;
  verification?: Record<string, unknown>;
  escrow?: Record<string, unknown>;
  dispute: { protocol: string; auto_file_threshold: number };
  signatures?: Record<string, string>;
  deliverable_hash?: string;
  delivered_at?: string;
  agreement_hash?: string;
}

export class Agreement {
  agreementId: string;
  asaVersion: string;
  createdAt: string;
  expiresAt: string;
  status: string;

  client: Identity | null;
  provider: Identity | null;
  evaluator: Identity | null;
  evaluatorType: string;

  service: ServiceSpec | null;
  qualityCriteria: QualityCriteria | null;
  verification: VerificationConfig | null;
  escrow: EscrowConfig | null;

  disputeProtocol: string;
  disputeAutoFileThreshold: number;

  clientSignature: string;
  providerSignature: string;

  deliverableHash: string;
  deliveredAt: string;
  agreementHash: string;

  constructor(opts: {
    agreementId?: string;
    asaVersion?: string;
    createdAt?: string;
    expiresAt?: string;
    status?: string;
    client?: Identity | null;
    provider?: Identity | null;
    evaluator?: Identity | null;
    evaluatorType?: string;
    service?: ServiceSpec | null;
    qualityCriteria?: QualityCriteria | null;
    verification?: VerificationConfig | null;
    escrow?: EscrowConfig | null;
    disputeProtocol?: string;
    disputeAutoFileThreshold?: number;
    clientSignature?: string;
    providerSignature?: string;
    deliverableHash?: string;
    deliveredAt?: string;
    agreementHash?: string;
  } = {}) {
    this.agreementId = opts.agreementId || `asa-${uuid().slice(0, 12)}`;
    this.asaVersion = opts.asaVersion || ASA_VERSION;
    this.createdAt = opts.createdAt || nowIso();
    this.expiresAt = opts.expiresAt || "";
    this.status = opts.status || "proposed";
    this.client = opts.client ?? null;
    this.provider = opts.provider ?? null;
    this.evaluator = opts.evaluator ?? null;
    this.evaluatorType = opts.evaluatorType || "agent_as_judge";
    this.service = opts.service ?? null;
    this.qualityCriteria = opts.qualityCriteria ?? null;
    this.verification = opts.verification ?? null;
    this.escrow = opts.escrow ?? null;
    this.disputeProtocol = opts.disputeProtocol || "ajp";
    this.disputeAutoFileThreshold = opts.disputeAutoFileThreshold ?? 60.0;
    this.clientSignature = opts.clientSignature || "";
    this.providerSignature = opts.providerSignature || "";
    this.deliverableHash = opts.deliverableHash || "";
    this.deliveredAt = opts.deliveredAt || "";
    this.agreementHash = opts.agreementHash || "";
  }

  computeHash(): string {
    const d = this._canonicalDict();
    this.agreementHash = hashDict(d);
    return this.agreementHash;
  }

  private _canonicalDict(): Record<string, unknown> {
    const d: Record<string, unknown> = {
      agreement_id: this.agreementId,
      asa_version: this.asaVersion,
      created_at: this.createdAt,
    };
    if (this.client) d.client = this.client.toDict();
    if (this.provider) d.provider = this.provider.toDict();
    if (this.service) d.service = this.service.toDict();
    if (this.qualityCriteria) d.quality_criteria = this.qualityCriteria.toDict();
    if (this.escrow) d.escrow = this.escrow.toDict();
    return d;
  }

  validate(): string[] {
    const errors: string[] = [];
    if (!this.client) errors.push("Agreement must have a client identity");
    if (!this.provider) errors.push("Agreement must have a provider identity");
    if (!this.service) errors.push("Agreement must specify a service");
    if (!this.qualityCriteria) {
      errors.push("Agreement must have quality criteria");
    } else {
      if (!this.qualityCriteria.dimensions.length) {
        errors.push("Quality criteria must include at least one dimension");
      } else {
        const totalWeight = this.qualityCriteria.dimensions.reduce((s, d) => s + d.weight, 0);
        if (Math.abs(totalWeight - 1.0) > 0.05) {
          errors.push(`Dimension weights should sum to ~1.0, got ${totalWeight.toFixed(2)}`);
        }
      }
    }
    if (!(AGREEMENT_STATUSES as readonly string[]).includes(this.status)) {
      errors.push(`Invalid status: ${this.status}`);
    }
    return errors;
  }

  isValid(): boolean {
    return this.validate().length === 0;
  }

  sign(party: "client" | "provider", signature: string): void {
    if (party === "client") {
      this.clientSignature = signature;
    } else if (party === "provider") {
      this.providerSignature = signature;
    }
    if (this.clientSignature && this.providerSignature) {
      this.status = "active";
      this.computeHash();
    }
  }

  deliver(deliverableHash: string): void {
    if (this.status !== "active") {
      throw new Error(`Cannot deliver in status '${this.status}', must be 'active'`);
    }
    this.deliverableHash = deliverableHash;
    this.deliveredAt = nowIso();
    this.status = "delivered";
  }

  markVerified(_passed: boolean): void {
    if (this.status !== "delivered") {
      throw new Error(`Cannot verify in status '${this.status}', must be 'delivered'`);
    }
    this.status = "verified";
  }

  close(): void {
    if (this.status !== "verified" && this.status !== "disputed") {
      throw new Error(`Cannot close in status '${this.status}'`);
    }
    this.status = "closed";
  }

  dispute(): void {
    if (this.status !== "verified" && this.status !== "delivered") {
      throw new Error(`Cannot dispute in status '${this.status}'`);
    }
    this.status = "disputed";
  }

  expire(): void {
    this.status = "expired";
  }

  reject(): void {
    if (this.status !== "proposed" && this.status !== "negotiating") {
      throw new Error(`Cannot reject in status '${this.status}'`);
    }
    this.status = "rejected";
  }

  toDict(): AgreementDict {
    const d: Record<string, unknown> = {
      asa_version: this.asaVersion,
      agreement_id: this.agreementId,
      created_at: this.createdAt,
      status: this.status,
    };
    if (this.expiresAt) d.expires_at = this.expiresAt;

    const parties: Record<string, unknown> = {};
    if (this.client) parties.client = this.client.toDict();
    if (this.provider) parties.provider = this.provider.toDict();
    if (this.evaluator) {
      const evDict = { ...this.evaluator.toDict(), type: this.evaluatorType };
      parties.evaluator = evDict;
    }
    if (Object.keys(parties).length > 0) d.parties = parties;

    if (this.service) d.service = this.service.toDict();
    if (this.qualityCriteria) d.quality_criteria = this.qualityCriteria.toDict();
    if (this.verification) d.verification = this.verification.toDict();
    if (this.escrow) d.escrow = this.escrow.toDict();

    d.dispute = {
      protocol: this.disputeProtocol,
      auto_file_threshold: this.disputeAutoFileThreshold,
    };

    const sigs: Record<string, string> = {};
    if (this.clientSignature) sigs.client = this.clientSignature;
    if (this.providerSignature) sigs.provider = this.providerSignature;
    if (Object.keys(sigs).length > 0) d.signatures = sigs;

    if (this.deliverableHash) d.deliverable_hash = this.deliverableHash;
    if (this.deliveredAt) d.delivered_at = this.deliveredAt;
    if (this.agreementHash) d.agreement_hash = this.agreementHash;

    return d as unknown as AgreementDict;
  }

  static fromDict(d: Record<string, unknown>): Agreement {
    const parties = (d.parties || {}) as Record<string, Record<string, unknown>>;
    const clientD = parties.client as Record<string, string> | undefined;
    const providerD = parties.provider as Record<string, string> | undefined;
    const evaluatorD = parties.evaluator as Record<string, string> | undefined;
    const sigs = (d.signatures || {}) as Record<string, string>;
    const dispute = (d.dispute || {}) as Record<string, unknown>;

    let evaluatorType = "agent_as_judge";
    let evaluatorForParse: Record<string, string> | undefined;
    if (evaluatorD) {
      evaluatorType = (evaluatorD.type as string) || "agent_as_judge";
      evaluatorForParse = { ...evaluatorD };
      delete evaluatorForParse.type;
    }

    return new Agreement({
      agreementId: (d.agreement_id as string) || "",
      asaVersion: (d.asa_version as string) || ASA_VERSION,
      createdAt: (d.created_at as string) || "",
      expiresAt: (d.expires_at as string) || "",
      status: (d.status as string) || "proposed",
      client: clientD ? Identity.fromDict(clientD as any) : null,
      provider: providerD ? Identity.fromDict(providerD as any) : null,
      evaluator: evaluatorForParse ? Identity.fromDict(evaluatorForParse as any) : null,
      evaluatorType,
      service: d.service ? ServiceSpec.fromDict(d.service as any) : null,
      qualityCriteria: d.quality_criteria ? QualityCriteria.fromDict(d.quality_criteria as any) : null,
      verification: d.verification ? VerificationConfig.fromDict(d.verification as any) : null,
      escrow: d.escrow ? EscrowConfig.fromDict(d.escrow as any) : null,
      disputeProtocol: (dispute.protocol as string) || "ajp",
      disputeAutoFileThreshold: (dispute.auto_file_threshold as number) ?? 60.0,
      clientSignature: sigs.client || "",
      providerSignature: sigs.provider || "",
      deliverableHash: (d.deliverable_hash as string) || "",
      deliveredAt: (d.delivered_at as string) || "",
      agreementHash: (d.agreement_hash as string) || "",
    });
  }

  toJson(indent = 2): string {
    return JSON.stringify(this.toDict(), null, indent);
  }

  static fromJson(s: string): Agreement {
    return Agreement.fromDict(JSON.parse(s));
  }
}
